### DISCUSSION TITLE
[Chapter 5] - Q & A

### DISCUSSION THREAD

**Discussion by Irisation23** [Study-2-Modern-Java-In-Action/Study-2-Modern-Java-in-Action#56] [CHAPTER 5 스트림 활용 - Q&A] [answered]
## 1. `findFirst()` 와 `findAny()`은 언제 사용할까?

- p169 참조

왜 findFirst()와 findAny()가 필요할까?
```
바로 병렬성 때문이다.
병렬 실행에서는 첫 번째 요소를 찾기 어렵다.

따라서, 요소의 반환 순서가 상관없다면 병렬 스트림에서는 제약이 적은 findAny()를 사용한다.
```

> Q1. 왜 병렬 실행에서는 첫 번째 요소를 찾기 어려울까?

> Q2. 왜 findFirst()보다 findAny()가 병렬 스트림에 제약이 적을까?

---

## 공통질문

현재 모던 자바 인 액션 책은 슬금 슬금 `병렬성`을 현시점에 끼어넣고 있다.
현재 `병렬성` 문제를 의구심 갖기 보다는 뒤로 Throw? 하는것이 좋을까?

**Comment by coalong** [comment, marked as answer]
### 공통질문!
p. 169 부분을 읽으면서 stream 의 병렬 처리 내부 동작을 알 수 없어서 이해하는데 더 난해했고, @bunsung92 님 질문대로 `병렬성` 에 깊게 의구심을 갖기 보다 뒤에 나올 내용을 생각하면서 Throw 하는 것이 좋다고 생각했습니다!
 


 Q1. 첫 번째 요소를 찾기 어렵다. -> 라는 문장을 
"순차적 처리일 때 첫 번째 요소를 찾는 것보다 병렬 처리일 때 `상대적으로` 첫 번째 요소를 찾는게 더 어렵다." 라고 이해했습니다.

 Q2. `findFirst` 는 순서에 의존하는 계산이라고 생각되고 `findAny` 순서에 의존하지 않는 계산이라고 생각했습니다.
그래서 병렬처리(Stream, 흐름을 쪼개고 다시 합치는 과정이 들어간...즉,  흐름==순서를 흐트려서 처리?)할 때는 `findFirst` 를 사용하기에 더 많은 비용이 든다고(제약이 있다) 이해했습니다.
